### PR TITLE
Migrate Snapshot repository version to IndexVersion

### PR DESCRIPTION
**Pull request by thecoop**
This changes the repository metadata version to IndexVersion, rather than use Version. In some cases, this also changes the json representation from a string to an int.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -364,7 +364,7 @@ public ClusterState execute(ClusterState currentState) {
                     repositoryData.resolveIndices(matchingIndices),
                     threadPool.absoluteTimeInMillis(),
                     repositoryData.getGenId(),
-                    minCompatibleVersion(currentState.nodes().getMinNodeVersion(), repositoryData, null)
+                    minCompatibleVersion(currentState.nodes().getMaxDataNodeCompatibleIndexVersion(), repositoryData, null)
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:367]: These are not identical - `getMinNodeVersion` considers all nodes, `getMaxDataNodeCompatibleIndexVersion` only considers data/master nodes. But snapshots are only relevant to data nodes, AFAIK.

File: server/src/test/java/org/elasticsearch/repositories/RepositoryDataTests.java [hunk 2]
```diff
@@ -111,7 +110,7 @@ public void testAddSnapshots() {
             newSnapshot,
             new RepositoryData.SnapshotDetails(
                 randomFrom(SnapshotState.SUCCESS, SnapshotState.PARTIAL, SnapshotState.FAILED),
-                randomFrom(IndexVersion.current(), Version.CURRENT.minimumCompatibilityVersion().indexVersion),
+                randomFrom(IndexVersion.current(), IndexVersion.MINIMUM_COMPATIBLE),
```
Comment (thecoop) [inline server/src/test/java/org/elasticsearch/repositories/RepositoryDataTests.java:113]: Not exactly the same, but still works for the tests

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 3]
```diff
@@ -816,10 +824,29 @@ public static RepositoryData snapshotsFromXContent(XContentParser parser, long g
...
+                    switch (parser.nextToken()) {
+                        case VALUE_STRING -> {
+                            Version v = Version.fromString(parser.text());
+                            assert v.before(Version.V_8_10_0);
+                            version = IndexVersion.fromId(v.id);
+                        }
+                        case VALUE_NUMBER -> {
+                            version = IndexVersion.fromId(parser.intValue());
+                        }
+                        default -> throw new ParsingException(
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:837]: There's a bug here, you would want to pass `parser.currentToken()` to the string formatting below, but that said:
I think we can just use `org.elasticsearch.common.xcontent.XContentParserUtils#throwUnknownToken` here, it gives us the same amount of information without blowing up the code size so much.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:367]: Jup this should be ok, only master and data nodes interact with the repository.

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 4]
```diff
@@ -816,10 +824,29 @@ public static RepositoryData snapshotsFromXContent(XContentParser parser, long g
                     indexMetaIdentifiers = parser.mapStrings();
                 }
                 case MIN_VERSION -> {
-                    XContentParserUtils.ensureExpectedToken(XContentParser.Token.VALUE_STRING, parser.nextToken(), parser);
-                    final Version version = Version.fromString(parser.text());
+                    IndexVersion version;
+                    switch (parser.nextToken()) {
+                        case VALUE_STRING -> {
+                            Version v = Version.fromString(parser.text());
+                            assert v.before(Version.V_8_10_0);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:831]: We could leave out this assertion and just use the same code for parsing the version here and in `parseSnapshots` below I think :) This code is complicated enough as it is (especially its BwC), would be nice to not duplicate that logic.

**Review by original-brownbear — COMMENTED** [review]
Looks just fine, just one issue and suggested fix in the parsing of RepositoryData.

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java (hunk 3 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:837]: Ah yes, didn't know about that method!

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java (hunk 4 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:831]: Pulled it out into a shared method

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 5]
```diff
@@ -980,6 +953,23 @@ private static void parseSnapshots(
...
+            }
+            case VALUE_NUMBER -> {
+                return IndexVersion.fromId(parser.intValue());
+            }
+            default -> {
+                XContentParserUtils.throwUnknownToken(parser.currentToken(), parser);
+                throw new AssertionError(); // never reached
+            }
+        }
+    }
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:971]: How about:

```java    
private static IndexVersion parseIndexVersion(XContentParser.Token token, XContentParser parser) throws IOException {
        if (token == XContentParser.Token.VALUE_NUMBER) {
            return IndexVersion.fromId(parser.intValue());
        } else {
            XContentParserUtils.ensureExpectedToken(XContentParser.Token.VALUE_STRING, token, parser);
            Version v = Version.fromString(parser.text());
            assert v.before(Version.V_8_10_0);
            return IndexVersion.fromId(v.id);
        }
    }
```

This should be faster (which does actually matter a little here) and avoids the confusing throwing of `AssertionError`?

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:971]: Hmm, or maybe change `ensureExpectedToken` to return an `Exception`, that can be used in a `throw` (even if the method itself throws directly), to keep the java compiler happy?

    switch {
        ...
        default -> throw XContentParserUtils.ensureExpectedToken(...);
    }

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:971]: Not sure that would be a good idea, it's definitely slower here I think and it's more code. The current form of that method I think is what compiles to the fastest code and is a convenient one-liner that can be used like an assertion.
Also that method is used 300+ times, we can't just change it :)

Comment (thecoop) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java]: OK, changed to suggestion

**Review by original-brownbear — APPROVED** [review]
LGTM thanks!

**Pull request merged**